File: src/app/model/PowerMeterData.ts
export class PowerMeterData {
    constructor(
        public timestamp: Date,
        public power: number,

        // If the sensor provides the power balance referenced to the left pedal, 
        // the power balance is calculated as [LeftPower/(LeftPower + RightPower)]*100 in units of percent
        public balance?: number,
        public accumulatedTorque?: number,
        public cumulativeCrankRevolutions?: number,
        public lastCrankEventTimestamp?: number  // Last crank event time Uint16 in 1/1024 second
    ) { }

    public diffCrankRotations(previous: PowerMeterData): number | null {
        return this.diff(this.cumulativeCrankRevolutions, previous.cumulativeCrankRevolutions, 0xFFFFFFFF);
    }


    public diffCrankTimeInSeconds(previous: PowerMeterData): number | null {
        const diff = this.diff(this.lastCrankEventTimestamp, previous.lastCrankEventTimestamp, 0xFFFFFFFF);
        if (diff === null) {
            return null;
        }
        return diff / 1024;
    }

    public getTotalPower(): number {
        if (this.balance == undefined || this.balance === null) {
            return this.power;
        }
        const balence = this.balance / 100;
        return this.power * (1 / balence);
    }

    private diff(biggest: number | undefined, smallest: number | undefined, maxValue: number): number | null {
        if (biggest === undefined || smallest === undefined) {
            return null;
        }
        if (biggest < smallest) {
            return (biggest + maxValue) - smallest;
        }
        return biggest - smallest;
    }
}

File: android/app/src/main/java/com/openbikecomputer/model/PowerMeterData.java
package com.openbikecomputer.model;

/**
 * Parsed Bluetooth cycling-power measurement. Ported from {@code PowerMeterData.ts}.
 */
public class PowerMeterData {
    /** Epoch milliseconds. */
    public long timestamp;
    public double power;

    /**
     * If the sensor provides the power balance referenced to the left pedal,
     * the power balance is {@code [LeftPower/(LeftPower + RightPower)]*100} in percent.
     */
    public Double balance;
    public Double accumulatedTorque;
    public Long cumulativeCrankRevolutions;
    /** Last crank event time, Uint16 in 1/1024 second. */
    public Long lastCrankEventTimestamp;

    public PowerMeterData(long timestamp, double power) {
        this.timestamp = timestamp;
        this.power = power;
    }

    public Double diffCrankRotations(PowerMeterData previous) {
        return diff(cumulativeCrankRevolutions, previous.cumulativeCrankRevolutions, 0xFFFFFFFFL);
    }

    public Double diffCrankTimeInSeconds(PowerMeterData previous) {
        Double d = diff(lastCrankEventTimestamp, previous.lastCrankEventTimestamp, 0xFFFFFFFFL);
        if (d == null) {
            return null;
        }
        return d / 1024.0;
    }

    public double getTotalPower() {
        if (balance == null || balance == 0.0) {
            return power * 2;
        }
        return power;
    }

    private static Double diff(Long biggest, Long smallest, long maxValue) {
        if (biggest == null || smallest == null) {
            return null;
        }
        if (biggest < smallest) {
            return (double) ((biggest + maxValue) - smallest);
        }
        return (double) (biggest - smallest);
    }
}
